<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="open_rviz"             default="true"/>
  <arg name="move_forward_only"     default="false"/>
  <arg name="mcl_map_file"          default="$(find raspicat_slam)/config/maps/iscas_museum_map.yaml"/>
  <arg name="navigation_map_file"   default="$(find raspicat_slam)/config/maps/iscas_museum_map.yaml"/>
  <arg name="navigation_map_topic"  default="/navigation_map"/>

  <arg name="navigation"            default="move_base" doc="navigation type [move_base, neonav]"/>

  <!-- Map server -->
  <node pkg="map_server" name="mcl_map_server" type="map_server" args="$(arg mcl_map_file)"/>

  <node pkg="map_server" name="move_base_map_server" type="map_server" args="$(arg navigation_map_file)">
    <remap from="map"         to="$(arg navigation_map_topic)"/>
    <remap from="static_map"  to="/navigation_static_map"/>
  </node>
  
  <!-- Amcl -->
  <include file="$(find raspicat_navigation)/launch/amcl.launch">
    <arg name="initial_pose_x" value="0.0"/>
    <arg name="initial_pose_y" value="0.0"/>
    <arg name="initial_pose_a" value="0.0"/>
  </include>

  <!-- Move base -->
  <group if="$(eval navigation=='move_base')">
    <include file="$(find raspicat_navigation)/launch/move_base.launch">
      <arg name="move_forward_only" value="$(arg move_forward_only)"/>
    </include>

    <node pkg="raspicat_navigation" name="lightsensors_to_laserscan_node" type="lightsensors_to_laserscan_node" 
          output="screen">
      <param name="left_side_usensor_frame_id"        value="left_side_usensor_link"/>
      <param name="left_front_usensor_frame_id"       value="left_front_usensor_link"/>
      <param name="right_front_usensor_frame_id"      value="right_front_usensor_link"/>
      <param name="right_side_usensor_frame_id"       value="right_side_usensor_link"/>
      <param name="usensor_max_threshold"             value="4000"/>
      <param name="usensor_min_threshold"             value="30"/>
      <param name="usensor_topic_receive_tolerance"   value="1.0"/>
    </node>
  </group>

  <!-- Neonavigation -->
  <group if="$(eval navigation=='neonav')">
    <include file="$(find raspicat_navigation)/launch/neonavigation.launch">
      <arg name="map_topic" value="$(arg navigation_map_topic)"/>
    </include>
  </group>

  <!-- Rviz -->
  <group if="$(eval open_rviz==true and navigation=='move_base')"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true"
      args="-d $(find raspicat_navigation)/config/rviz/move_base.rviz">
    </node>
  </group>
  
  <group if="$(eval open_rviz==true and navigation=='neonav')"> 
    <node pkg="rviz" type="rviz" name="rviz" required="true"
      args="-d $(find raspicat_navigation)/config/rviz/neonav.rviz">
    </node>
  </group>
</launch>